src/main/java/org/reactome/addlinks/dataretrieval/ensembl/EnsemblBiomartRetriever.java

```diff
@@ -0,0 +1,232 @@
+     * If/When this exception is thrown for this species, it can be ignored. Check the logs from previous runs to confirm.
+     * @param queryString - String, URL/XML string that will be used to query BioMart.
+     * @param retryCount - int, Denotes how many times this query has been tried with BioMart.
+     * @return - Set<String>, All lines of successful BioMart response.
+     * @throws IOException - Thrown by HttpURLConnection, BufferedReader, URL classes.
+     * @throws InterruptedException - Thrown if Sleep is interrupted when waiting to retry BioMart query.
+     * @throws BioMartQueryException - Thrown if BioMart query doesn't match any existing data in their database.
+     * @throws HttpException - Thrown when the Http request to BioMart returns a non-200 response.
+     */
+    private Set<String> queryBiomart(String queryString, int retryCount) throws IOException, InterruptedException, BioMartQueryException, HttpException {
+        final int MAX_QUERY_RETRIES = 5;
+        // Create connection to Biomart URL for each species, retrieving a list of microarray probe types, if available.
+        URL biomartUrlWithSpecies = new URL(queryString);
+        HttpURLConnection biomartConnection = (HttpURLConnection) biomartUrlWithSpecies.openConnection();
+        if (biomartConnection.getResponseCode() == HttpStatus.SC_OK) {
+            Set<String> biomartResponseLines = new HashSet<>();
+            BufferedReader br = new BufferedReader(new InputStreamReader(biomartConnection.getInputStream()));
+            String line;
+            while ((line = br.readLine()) != null) {
+                // Biomart still responds with a 200, even if no data exists. For now, we handle it by
+                // checking the returned content for the string 'ERROR'. It will retry up to 5 times, with a 10 second delay.
+                if (line.contains("ERROR") ) {
+                    retryCount++;
+                    if (retryCount < MAX_QUERY_RETRIES) {
+                        return retryQuery(queryString, retryCount);
+                    }
+                    // The data does not exist in BioMart for a few species. Frog (X. tropicalis) has UniProt data but not Microarray data;
+                    // Yeast (S. cerevisiae) has Uniprot-SwissProt data but not UniProt-TrEMBL data; Yeast (S. pombe), P. falciparum
+                    // and D. discoideum don't have UniProt or Microarray data, all at time of writing (January 2020).
+//                    throw new Exception("Biomart query failed with message: " + line + "\nThis can happen without issue for certain species (X. tropicalis, S. pombe, S. cerevisiae, P. falciparum) because the data doesn't exist in BioMart");
+                    throw new BioMartQueryException(line +
+                            "\nThis can happen without issue for certain species (X. tropicalis, S. pombe, S. cerevisiae, P. falciparum) " +
+                            "because the data doesn't exist in BioMart");
```

Comment:
It might be clearer to invert this check and have the exception thrown `if >= MAX_QUERY_RETRIES` and then call `retryQuery`

Reply:
Done

Reply:
Hah. I find it the exact opposite! :)

...`retryCount` is what's modified, and in my head, I always compare that to some number, not compare some number to the variable. So I find the _new_ code to be weird and awkward to read. Overall, this sort of thing is potentially highly subjective. One person's clarity is another's... unclarity. Oh well, it's done now. _shrug_